src/librustc/hir/mod.rs

```diff
@@ -538,6 +539,7 @@ pub enum GenericParamKind {
 #[derive(Clone, RustcEncodable, RustcDecodable, Debug)]
 pub struct GenericParam {
     pub id: NodeId,
```

Comment:
Can we avoid having `NodeId` at all? Then this would be a clear win. Have you looked at the remaining uses of `id` for `GenericParam` and `Lifetime`?

Reply:
I have, and there were quite a few. I can try giving it a shot :+1:.

Reply:
@varkor Actually this is super complex; `NodeId` is still used in lots of places and just removing it causes lots of breakage, making it hard to debug. I can see that most `Hir` objects still point to `NodeId`, probably for the same reason. I have a suggestion: I will try to migrate methods working on `NodeId` to `HirId` in following PRs; at one point the compiler should just tell us that the `NodeId` field is not used anymore - that approach should be a lot cleaner.